### PR TITLE
Updating to `numpy` 2.0.0

### LINKED ISSUE #453: Explore move to numpy 2.0
**Is your feature request related to a problem? Please describe.**

Numpy 2.0 is upon us. It looks like xarray has updated to support it and `ruff` provides a migration rule.

**Describe the solution you'd like**

Create a branch, update numpy requirement and add the `NPY201` ruff rule.

### PR DESCRIPTION
**Pull request by davidorme**
# Description

This PR sets out the changes needed to move to `numpy` 2.0.0. Its remarkably clean:

* Need to bump versions of `numpy` (natch) but also `xarray` and `pint` which are now `numpy` 2.0.0 compatible.
* Add the `NPY201` rule for our `ruff` config.
* A couple of replacements in the code: `np.infty` to `np.inf` and `np.row_stack` to `np.vstack`

Fixes #453

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -125,8 +125,10 @@ class MyReferenceStyle(AuthorYearReferenceStyle):
     ("py:class", "dataclasses.InitVar"),
     ("py:class", "Quantity"),
     ("py:class", "numpy._typing._array_like._ScalarType_co"),
-    # TODO - Delete this once Vivienne has merged this feature into develop
-    ("py:class", "virtual_ecosystem.models.abiotic.energy_balance.EnergyBalance"),
+    # God only knows why this is needed. We don't refer to pint.util.Quantity and it
+    # isn't in the pint objects.inv, so why the hell is intersphinx trying to build
+    # references to it.
+    ("py:class", "pint.util.Quantity"),
```
Comment (alexdewar) [inline docs/source/conf.py:131]: Wait, should this be `pint.Quantity`, as you've written it below? I'm confused

**Review by alexdewar — COMMENTED** [review]
In the description you mention a few changes to the code, but it doesn't seem that there are any numpy-related changes in this PR... Have I missed something?

File: docs/source/conf.py (hunk 1 above)
Comment (dalonsoa) [reply docs/source/conf.py:131]: I'm also confused. There's also a `Quantity` up in line 126, which I'm not sure if it is the same quantity.

**Review by dalonsoa — APPROVED** [review]
Like Alex says in his comment, this looks ok, but it seems there are more changes related to `pint` that `numpy`. None of the `infty`-related ones are here, for exampel.

File: docs/source/conf.py (hunk 1 above)
Comment (davidorme) [reply docs/source/conf.py:131]: @alexdewar and @dalonsoa

I should have made this clearer. Asides from updating the `pyproject.toml` references and adding the Ruff rule, there were basically only about 8 lines that needed changing (`np.infty`, `np.float_` and `np.column_stack`) and it looks like all those changes have been made in branches that have already merged to `develop`.

The `pint` stuff is a pain - we needed to update `pint` as our current version wasn't `numpy 2.0` compliant, but there is something off with the `pint` documentation and `objects.inv` that means that whatever we seem to do, we end up with `sphinx` choking on this mysterious `pint.util.Quantity`. There may be a fix - I'm similarly puzzled by why we need so many `numpy` nitpick ignores.

**Pull request merged**